R/tm_g_heat_bygrade.R

```diff
@@ -352,62 +350,75 @@ srv_g_heatmap_bygrade <- function(id,
-
       ADSL <- data[[sl_dataname]]() # nolint
       ADEX <- data[[ex_dataname]]() # nolint
       ADAE <- data[[ae_dataname]]() # nolint
+      if (isTRUE(input$plot_cm)) {
+        ADCM <- data[[cm_dataname]]() # nolint
+      }
 
-      validate(need(nrow(ADSL) > 0, "Please select at least one subject"))
-      validate(need(
-        input$ongo_var %in% names(ADEX),
-        paste("Study Ongoing Status must be a variable in", ex_dataname, sep = " ")
-      ))
+      teal::validate_has_data(ADSL, min_nrow = 0, msg = sprintf("%s contains no data", sl_dataname))
 
-      validate(need(
-        checkmate::test_logical(ADEX[[input$ongo_var]], min.len = 1, any.missing = FALSE),
+      iv <- shinyvalidate::InputValidator$new()
+      iv$add_rule("id_var", shinyvalidate::sv_required(
+        message = "ID Variable is required"
+      ))
+      iv$add_rule("visit_var", shinyvalidate::sv_required(
+        message = "Visit Variable is required"
+      ))
+      iv$add_rule("ongo_var", shinyvalidate::sv_required(
+        message = "Study Ongoing Status Variable is required"
+      ))
+      iv$add_rule("ongo_var", shinyvalidate::sv_in_set(
+        set = names(ADEX),
+        message_fmt = sprintf("Study Ongoing Status must be a variable in %s", ex_dataname)
+      ))
+      iv$add_rule("ongo_var", ~ if (!is.logical(ADEX[[req(.)]])) {
         "Study Ongoing Status must be a logical variable"
+      })
+      iv$add_rule("anno_var", shinyvalidate::sv_required(
+        message = "Annotation Variables is required"
       ))
-
-      validate(need(
-        all(anno_var %in% names(ADSL)),
-        paste("Please only select annotation variable(s) in", sl_dataname, sep = " ")
+      iv$add_rule("anno_var", ~ if (length(.) > 2L) {
+        "No more than two Annotation Variables are allowed"
+      })
+      iv$add_rule("anno_var", shinyvalidate::sv_in_set(
+        set = names(ADSL),
+        message_fmt = sprintf("Study Ongoing Status must be a variable in %s", sl_dataname)
+      ))
+      iv$add_rule("anno_var", ~ if (isTRUE(input$id_var %in% .)) {
+        sprintf("Deselect %s in Annotation Variables", input$id_var)
+      })
+      iv$add_rule("heat_var", shinyvalidate::sv_required(
+        message = "Heat Variable is required"
       ))
+      iv$enable()
 
-      validate(need(
-        !(input$id_var %in% anno_var),
-        paste("Please de-select", input$id_var, "in annotation variable(s)", sep = " ")
+      iv_cm <- shinyvalidate::InputValidator$new()
+      iv_cm$condition(~ isTRUE(input$plot_cm))
+      iv_cm$add_rule("conmed_var", shinyvalidate::sv_required(
+        message = "Conmed Variable is required"
+      ))
+      iv_cm$add_rule("conmed_var", shinyvalidate::sv_in_set(
+        set = names(ADCM),
+        message_fmt = sprintf("Conmed Variable must be a variable in %s", cm_dataname)
       ))
+      iv_cm$add_rule("conmed_var", ~ if (!is.factor(ADCM[[.]])) {
+        "Study Ongoing Status must be a factor variable"
+      })
```

Comment:
Something is wrong with `conmed_var` that the validation is not rendered in the main plot space:
![Screenshot from 2022-12-12 09-41-00](https://user-images.githubusercontent.com/86970066/207001022-bd5c333f-35e6-42d5-8175-5987663c0de7.png)